### PR TITLE
fix(pages): publish qualified FPE snapshots

### PR DESCRIPTION
**Pull request by jasonqinzhou**
## Why and what changed

FPE refreshes uploaded new CSVs, but Pages deployed the old committed snapshot. Every public Pages deployment now selects the newest qualified tested source in main history, including deployments triggered by successful main FPE/Nightly runs. The public page shows snapshot time, tested SHA, and the producing CI link; web artifacts are retained for 90 days.

The selector independently checks workflow identity and actual tested provenance, validates the qualification and all CSV identities/shards, and copies only indexed data into trusted repository HTML. Manual dispatches can test an `expected_sha` different from the workflow event SHA. The run attempt must be a positive integer; malformed or missing attempt metadata stops publication. A partial rerun that leaves the winning qualified snapshot's latest attempt unsuccessful stops deployment before writing data. Strict JSON validation also rejects exponent overflow before it can become browser-invalid `Infinity`.

## Review map

- Risk: **High** — public deployment workflow and artifact trust boundary.
- Start with `scripts/prepare_fpe_pages.py`, then `.github/workflows/pages.yml`, `scripts/build_pages_site.py`, and `tests/test_pages_site.py`.
- Serialized change: additive `snapshot` metadata in the deployed FPE index; existing CSV columns remain compatible. PR previews use repository data.
- Availability tradeoff: an expired artifact has unknown tested provenance and can stop publication unless a higher-ID validated current-HEAD snapshot already proves later candidates cannot win. An expired historical run can therefore block an older-than-HEAD snapshot; refresh FPE at current main to restore publication.
- Rollback: revert the PR and redeploy Pages to restore historical repository data. Production deployment remains unverified until merge and Pages deployment.

## Evidence

Validated and independently reviewed implementation commit: `c69adb8b60221f79bc3cc1879a8bd57b8f61a1e3`.

- **Local suites:** **167 passed**, including **39 Pages tests**, using `uv run --no-project --python 3.12 --with pytest==8.4.2 --with pyyaml==6.0.3 python -m pytest -c /dev/null tests/test_pages_site.py tests/test_ci_workflow_contracts.py tests/test_ci_qualification.py -q -p no:cacheprovider -p no:timeout --override-ini='addopts=' --override-ini='filterwarnings='`. These are publication and workflow/qualification tests, not GPU accuracy tests.
- **Independent review:** final whole-branch Standards and Spec re-reviews of this exact commit are clean, with zero remaining or deferred findings. Both independently ran all 167 tests and verified the JSON failure before/after through the public consumer. Earlier checks covered seven real-producer lifecycle scenarios, 11 additional edge cases, and an independent 216-case selection/status oracle using real Git ancestry.
- **Real artifact and browser:** selected artifact `10371902391` from [qualified main FPE run 34900977951](https://github.com/ai-dynamo/aisimulate/actions/runs/34900977951), tested source `be8f69749589aea5a376083436e054d75810d379`: **4,134 rows, 78 models, 10 systems**. Live retrieval/build/browser validation succeeded during the fix; the final commit also replayed the recorded real API responses/archive, built 31 public files, and rendered 78 model rows with correct provenance and no JavaScript errors. Repository preview rendered 71 historical rows. A repeat live download during validation encountered transport timeouts; the final replay used the previously downloaded real archive.
- **Static checks:** Ruff lint/format and whitespace checks passed on the corrections. The full branch also passed actionlint, SPDX/legal-file checks, and CODEOWNERS validation; hosted evidence below is tied to the final commit.
- **Current main:** the prior correction at `01c00997` was combined with `6757084d20edcd800a98a1efd320f94f53f0dfd4` in a disposable worktree: conflict-free, no overlapping changed paths, 164 tests passed, and 31 Pages files built. Hosted checks below validate the final PR head separately.
- **Boundary coverage:** tested SHA/event SHA mismatch, historical dispatch ordering, partial retries, artifact-ID ties, expiry, pagination, off-main sources, malformed/off-main qualification before fallback, mixed CSV identities, missing/unsafe data, missing or invalid run-attempt metadata, shard completeness, strict counts/statuses, duplicate JSON/CSV identities, malformed records, numeric constants/overflow at all depths, finite metadata through JavaScript parsing, and trusted site content preservation.

### Hosted checks on the final commit

- **Fast CI:** [Passed](https://github.com/ai-dynamo/aisimulate/actions/runs/34951808231).
- **Pages preview:** [Passed; deployment correctly skipped for PR](https://github.com/ai-dynamo/aisimulate/actions/runs/34951808100).
- **DCO:** [Passed](https://github.com/ai-dynamo/aisimulate/runs/104324276724).
- **CODEOWNERS integrity:** [Passed](https://github.com/ai-dynamo/aisimulate/actions/runs/34951808205).
- **CodeRabbit:** [Passed; no actionable findings in final review](https://github.com/ai-dynamo/aisimulate/pull/209#issuecomment-5671783289).
- **Full CI:** [Passed; 56 jobs succeeded, 1 expected staging skip](https://github.com/ai-dynamo/aisimulate/actions/runs/34952807269).

Fast CI Success and Full CI Success passed on c69adb8b, together with Pages preview, DCO, and CODEOWNERS integrity. CodeRabbit completed the final review with no actionable findings; both independent Codex review axes are clean. All review conversations are resolved. Full CI ran after those reviews with expected_sha=c69adb8b60221f79bc3cc1879a8bd57b8f61a1e3: 56 jobs passed, with only the expected manual-run application-wheel staging job skipped. GitHub still reports REVIEW_REQUIRED, so applicable human/CODEOWNER approvals remain outstanding. The PR is open and unmerged.

## Fixed review findings

- [Tested source differs from dispatch revision](https://github.com/ai-dynamo/aisimulate/pull/209#discussion_r4012530877): fixed in `d4b315c44edf03f5b632cf4d5db5aa1be7717fc6`.
- [Partial Nightly retry can roll back the published snapshot](https://github.com/ai-dynamo/aisimulate/pull/209#discussion_r4012530888): fixed in `d4b315c44edf03f5b632cf4d5db5aa1be7717fc6`.
- [JSON exponent overflow breaks the browser index](https://github.com/ai-dynamo/aisimulate/pull/209#discussion_r4012831881): fixed in `01c009971c796d50d6b3b0ff2c1c612c8b9db559`.
- [Missing or malformed run attempt can select older data](https://github.com/ai-dynamo/aisimulate/pull/209#discussion_r4013295724): fixed in `c69adb8b60221f79bc3cc1879a8bd57b8f61a1e3`.

## Modeling or data provenance

This changes publication and provenance display. The data remains strict-native op-level FPE coverage with qualification and row source identities preserved. It does not establish end-to-end serving or silicon accuracy. Estimator calculations and measured datasets are unchanged.

## Tracking

- Closes: none.
- Related: stale FPE website snapshot; [qualified source run](https://github.com/ai-dynamo/aisimulate/actions/runs/34900977951).

### CODE REVIEW

File: scripts/prepare_fpe_pages.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+            or report.get("qualification") != QUALIFICATION
+            or report.get("source_sha") != source_sha
+            or not re.fullmatch(r"[0-9a-f]{64}", str(report.get("wheel_sha256", "")))
+            or not isinstance(report.get("shard_count"), int)
+            or report["shard_count"] <= 0
+            or not isinstance(report.get("required_probe_count"), int)
+            or report["required_probe_count"] <= 0
+        ):
+            raise ValueError("invalid FPE qualification or mismatched source commit")
+        statuses = report.get("status_counts", {})
```
Comment (jasonqinzhou) [inline scripts/prepare_fpe_pages.py:83]: [P2] Validate all numeric qualification fields before publication. I reproduced acceptance of schema_version=true and status_counts.PASS values -1, true, and "4". These values are outside the producer contract but satisfy the current equality/truthiness checks, so malformed manifests can be published. Require an integer schema version and a status-count object containing nonnegative integer counts, with a positive PASS count.

File: scripts/prepare_fpe_pages.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        report = json.loads(bundle.read("fpe-qualification.json"))
+        if (
+            report.get("schema_version") != 1
+            or report.get("qualification") != QUALIFICATION
+            or report.get("source_sha") != source_sha
+            or not re.fullmatch(r"[0-9a-f]{64}", str(report.get("wheel_sha256", "")))
+            or not isinstance(report.get("shard_count"), int)
+            or report["shard_count"] <= 0
+            or not isinstance(report.get("required_probe_count"), int)
+            or report["required_probe_count"] <= 0
```
[the remark this answers is not in the document]
Comment (jasonqinzhou) [reply scripts/prepare_fpe_pages.py]: Fixed in 3191e56f3e3da6f09cbc2a6343e74333e14a404f. Qualification schema/count fields now require exact integers; status counts must be a JSON object with nonnegative integer values. Regression tests reject booleans, floats, strings, nulls, and negative values while accepting zero nonpassing counts. The Pages, workflow, and qualification suites pass (147 tests).

File: scripts/prepare_fpe_pages.py (hunk 1 above)
Comment (jasonqinzhou) [reply scripts/prepare_fpe_pages.py:83]: Fixed in 3191e56f3e3da6f09cbc2a6343e74333e14a404f. Qualification schema/count fields now require exact integers; status counts must be a JSON object with nonnegative integer values. Regression tests reject booleans, floats, strings, nulls, and negative values while accepting zero nonpassing counts. The Pages, workflow, and qualification suites pass (147 tests).

**Review by jasonqinzhou — COMMENTED** [review]
Codex self-review completed for `3191e56f3e3da6f09cbc2a6343e74333e14a404f` against `main` at `be8f69749589aea5a376083436e054d75810d379`.

Reviewed the artifact producer, workflow/run filtering, commit ordering, qualification and CSV validation, allowlisted site build, provenance UI, and failure/rollback paths. No remaining actionable issue found. The CodeRabbit boolean-count issue and the additional schema/status-count issue are fixed, regression-tested, and resolved.

Validation: 147 local Pages/workflow/qualification tests pass; Ruff, actionlint, SPDX/legal-file checks pass. The final head also successfully selected and built the real qualified main FPE artifact (4,134 rows, 78 models). Current-head Fast CI, Pages build, DCO, and CODEOWNERS pass.

The current-head CodeRabbit review is still running. Full CI will be dispatched only after that review completes, following REVIEW.md. This is a technical self-review, not an approval or merge-readiness claim; production deployment is verified after merge.

File: scripts/prepare_fpe_pages.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+def qualified_files(archive: bytes, source_sha: str) -> dict[str, bytes] | None:
+    """Read only qualified CSVs; old artifacts without a qualification are ineligible."""
+    with zipfile.ZipFile(io.BytesIO(archive)) as bundle:
+        names = bundle.namelist()
+        if len(names) != len(set(names)):
+            raise ValueError("duplicate archive entries")
+        if "fpe-qualification.json" not in names:
+            return None
+        report = json.loads(bundle.read("fpe-qualification.json"))
```
[the remark this answers is not in the document]
Comment (jasonqinzhou) [reply scripts/prepare_fpe_pages.py]: Fixed in 969f2684a0189c6e6710f36fbdfc098ec2375dc4. Both artifact JSON documents now use a strict decoder that rejects duplicate members at any object depth and non-JSON NaN/Infinity constants. Regression tests cover duplicate schema/status-count fields, nested duplicate PASS keys, and a duplicate files member in index.json. All 149 local Pages/workflow/qualification tests pass.

**Review by jasonqinzhou — COMMENTED** [review]
Codex self-review completed for `969f2684a0189c6e6710f36fbdfc098ec2375dc4` against `main` at `be8f69749589aea5a376083436e054d75810d379`.

Reviewed the artifact producer, workflow/run filtering, commit ordering, qualification and CSV validation, allowlisted site build, provenance UI, and failure/rollback paths. No remaining actionable issue found. The numeric-validation and duplicate-JSON findings are fixed and regression-tested.

Validation: 149 local Pages/workflow/qualification tests pass; Ruff, actionlint, SPDX/legal-file checks pass. The final head also successfully selected and built the real qualified main FPE artifact (4,134 rows, 78 models). Refreshed hosted checks are pending for this new commit.

The current-head CodeRabbit review is still running. Full CI will be dispatched only after that review completes, following REVIEW.md. This is a technical self-review, not an approval or merge-readiness claim; production deployment is verified after merge.

File: scripts/prepare_fpe_pages.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        files = index.get("files") if isinstance(index, dict) else None
+        if not isinstance(files, list) or not files or len(set(files)) != len(files):
+            raise ValueError("empty or duplicate FPE dataset index")
+        selected = {"index.json": json.dumps(index).encode()}
+        shards = set()
+        for filename in files:
+            if not isinstance(filename, str) or not re.fullmatch(r"[a-z0-9_]+\.csv", filename):
+                raise ValueError("unsafe FPE dataset filename")
+            data = bundle.read(DATA_PREFIX + filename)
+            rows = list(csv.DictReader(io.StringIO(data.decode("utf-8"))))
```
Comment (jasonqinzhou) [inline scripts/prepare_fpe_pages.py]: [P2] Reject ambiguous CSV records before publication. I reproduced acceptance of a CSV with duplicate Status headers containing FAIL then PASS, and of a repeated capability row. DictReader silently keeps the last duplicate field, and repeated identities can change the displayed result/counts. Validate unique headers, strict record shape/quoting, and one row per model/architecture/system/backend/version capability before accepting the dataset.

File: scripts/prepare_fpe_pages.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+            or type(report.get("shard_count")) is not int
+            or report["shard_count"] <= 0
+            or type(report.get("required_probe_count")) is not int
+            or report["required_probe_count"] <= 0
+        ):
+            raise ValueError("invalid FPE qualification or mismatched source commit")
+        statuses = report.get("status_counts", {})
+        if not isinstance(statuses, dict) or any(type(count) is not int or count < 0 for count in statuses.values()):
+            raise ValueError("FPE qualification status counts must be nonnegative integers")
+        if not statuses.get("PASS", 0) or statuses.get("BUILD_FAILED", 0) or statuses.get("QUERY_FAILED", 0):
```
[the remark this answers is not in the document]
Comment (jasonqinzhou) [reply scripts/prepare_fpe_pages.py:88]: Fixed in 48ddae1921ff918b0874edb8bbfa7199d0eb463b. Unknown status-count keys are rejected against the qualifier status set, with a producer/consumer consistency test. Whitespace-variant failure keys and other unknown keys are covered.

File: scripts/prepare_fpe_pages.py (hunk 4 above)
Comment (jasonqinzhou) [reply scripts/prepare_fpe_pages.py]: Fixed in 48ddae1921ff918b0874edb8bbfa7199d0eb463b. CSV decoding now enforces unique nonempty headers, strict quoting, consistent record widths, and unique capability identities. Reproduced duplicate-header and duplicate-row cases now fail closed. All 152 local Pages/workflow/qualification tests pass, and the real qualified main artifact still builds successfully.

**Review by jasonqinzhou — COMMENTED** [review]
Codex self-review completed for `48ddae1921ff918b0874edb8bbfa7199d0eb463b` against `main` at `be8f69749589aea5a376083436e054d75810d379`.

Reviewed the artifact producer, workflow/run filtering, commit ordering, qualification and CSV validation, allowlisted site build, provenance UI, and failure/rollback paths. No remaining actionable issue found. The numeric, JSON, status-key, and CSV-identity findings are fixed and regression-tested.

Validation: 152 local Pages/workflow/qualification tests pass; Ruff, actionlint, SPDX/legal-file checks pass. The final head also successfully selected and built the real qualified main FPE artifact (4,134 rows, 78 models). Refreshed hosted checks are pending for this new commit.

The current-head CodeRabbit review is still running. Full CI will be dispatched only after that review completes, following REVIEW.md. This is a technical self-review, not an approval or merge-readiness claim; production deployment is verified after merge.

**Review by jasonqinzhou — COMMENTED** [review]
Codex final readiness review: `48ddae1921ff918b0874edb8bbfa7199d0eb463b`. No remaining actionable finding; all five validated findings are fixed and all review threads are resolved. CodeRabbit confirms complete coverage of this same head.

[Full CI passed](https://github.com/ai-dynamo/aisimulate/actions/runs/34910835366), including `Full CI Success` (56 passing jobs; only push-only staging skipped). Fast CI, Pages preview build, DCO, and CODEOWNERS also passed. The PR description contains local test and real qualified-artifact evidence.

Current `main` is now `e46be717175acf06bdbbdeadb7aaf9bb2afdae8d`. I reviewed its nonoverlapping CLI changes and verified the combined tree in a disposable worktree: no merge conflicts, 152 tests passed, 31 public Pages files built, and whitespace checks passed. The PR head remains unchanged.

Ready for human review. Human/CODEOWNER approval and post-merge production deployment are still outstanding.

**Review by Arsene12358 — COMMENTED** [review]
Independent review of `48ddae1921ff918b0874edb8bbfa7199d0eb463b`: two reproduced merge blockers, one source-identity contract defect and one partial-rerun rollback. The 152 existing focused tests, real qualified-artifact build (4,134 rows / 78 models), and browser checks pass; the lifecycle cases below are missing from that coverage.

File: scripts/prepare_fpe_pages.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+            or identity.get("head_branch") != "main"
+            or identity.get("head_sha") not in ranks
+        ):
+            continue
+        candidates.append(artifact)
+    candidates.sort(key=lambda a: (ranks[a["workflow_run"]["head_sha"]], -a["id"]))
+    for artifact in candidates:
+        run_id = artifact["workflow_run"]["id"]
+        run = json.loads(api(repository, f"actions/runs/{run_id}"))
+        source_sha = artifact["workflow_run"]["head_sha"]
```
Comment (Arsene12358) [inline scripts/prepare_fpe_pages.py]: **[P2] Distinguish the tested source SHA from the dispatch revision.**

The existing producer checks out `inputs.expected_sha`; `generate_fpe_support_matrix.py:29–40` explicitly records that checkout because it can differ from the event SHA. A main-ref dispatch at A targeting an older main commit B is valid, including when main advances between resolving its SHA and starting the dispatch. This line instead uses A as the tested source, so `qualified_files()` rejects the valid B manifest and Pages aborts before considering the retained qualified snapshot.

I reproduced this with real main ancestors `be8f6974` / `9ad041ea`, the actual qualifier and CSV producer, and simulated Actions metadata: both archives qualify for their tested sources, but Pages raises `ValueError: invalid FPE qualification or mismatched source commit`.

| Current | Required |
| --- | --- |
| Rank and validate against the workflow dispatch SHA. | Keep workflow identity separate; validate and rank by the qualified tested source, with main ancestry checks. Alternatively align the producer contract before producing artifacts. |

Add coverage for an explicit older-main target and a dispatch-time branch advance.

File: scripts/prepare_fpe_pages.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+        if (
+            run.get("path") not in WORKFLOWS
+            or run.get("head_repository", {}).get("full_name") != repository
+            or run.get("head_branch") != "main"
+            or run.get("head_sha") != source_sha
+            or run.get("event") not in {"schedule", "workflow_dispatch"}
+            or run.get("status") != "completed"
+            or run.get("conclusion") != "success"
+        ):
+            continue
```
Comment (Arsene12358) [inline scripts/prepare_fpe_pages.py]: **[P1] Do not roll back a retained snapshot when an unrelated Nightly job is retried.**

`actions/runs/{run_id}` describes the latest attempt. After a successful Nightly publishes source A, rerunning only `stage-artifactory` retains its qualified FPE artifact, but the run becomes waiting/in-progress or can fail. A concurrent documentation deployment then takes this `continue` and publishes older source B. The FPE artifact that qualified A has not changed.

Reproduced using the actual qualifier, CSV producer, selector, and final Pages builder with simulated lifecycle responses: baseline publishes `be8f6974`; attempt 2 states `waiting`, `in_progress`, and `completed/failure` each produce a public `index.json` for older `9ad041ea`. This is local evidence, not an observed production rollback.

| Current | Required |
| --- | --- |
| Skip the newer artifact based on the latest attempt and deploy older data. | Verify the artifact's successful producing attempt, or stop deployment to preserve the existing site when that cannot be established. |

Add a staging-only partial-rerun regression test. GitHub documents [stable run IDs across attempts](https://docs.github.com/en/actions/reference/workflows-and-actions/variables) and [artifact retention on partial reruns](https://github.com/orgs/community/discussions/45415).

File: scripts/prepare_fpe_pages.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+        for key, value in pairs:
+            if key in result:
+                raise ValueError(f"duplicate JSON member: {key}")
+            result[key] = value
+        return result
+
+    def reject_constant(value):
+        raise ValueError(f"invalid JSON constant: {value}")
+
+    return json.loads(data, object_pairs_hook=unique_object, parse_constant=reject_constant)
```
Comment (Arsene12358) [inline scripts/prepare_fpe_pages.py]: [P2] Reject floating-point overflow before reserializing artifact JSON

`parse_constant` rejects literal `NaN`/`Infinity` but is not called for numeric overflow such as `1e999`. An otherwise qualified artifact index containing `{"files":["b200_sxm.csv"],"extra":1e999}` is accepted, then `json.dumps()` writes the extra value as bare `Infinity`. I reproduced this through `prepare()` and `build_site()`; JavaScript `JSON.parse()` rejects the deployed index, so the matrix cannot load. The existing producer currently writes only `files`; this is a malformed-artifact validation gap, not a claim that normal production artifacts contain the value. Reject non-finite parsed floats before any output and add a regression with an older valid fallback to ensure malformed data cannot silently select older content.

File: scripts/prepare_fpe_pages.py (hunk 6 above)
Comment (Arsene12358) [reply scripts/prepare_fpe_pages.py]: Fixed in d4b315c44edf03f5b632cf4d5db5aa1be7717fc6. The selector independently checks workflow identity, validates the manifest and every CSV row against the actual tested SHA, then ranks that SHA within current main history. A later historical dispatch cannot displace a newer tested source. Regression coverage also follows the selected data through the final site builder. Independent whole-branch Standards and Spec re-reviews are clean at final head 01c009971c796d50d6b3b0ff2c1c612c8b9db559; 164 focused tests passed.

File: scripts/prepare_fpe_pages.py (hunk 7 above)
Comment (Arsene12358) [reply scripts/prepare_fpe_pages.py]: Fixed in d4b315c44edf03f5b632cf4d5db5aa1be7717fc6. Retained qualified artifacts from retried runs participate in tested-source ordering. If the winning snapshot belongs to a retry whose latest attempt is not successful, preparation raises before writing deployment data, preserving the existing site. An older retry does not block a newer successful snapshot. Tests cover waiting, running, failed, and cancelled retries through producer-to-site lifecycle reproductions. Independent final re-reviews are clean at 01c009971c796d50d6b3b0ff2c1c612c8b9db559; 164 focused tests passed.

File: scripts/prepare_fpe_pages.py (hunk 8 above)
Comment (Arsene12358) [reply scripts/prepare_fpe_pages.py]: Fixed in 01c009971c796d50d6b3b0ff2c1c612c8b9db559. The strict parser now rejects non-finite results from numeric exponent parsing at any nesting depth. Eight negative cases cover both signs in the qualification manifest and index, with an older valid fallback and an assertion that no output directory is created. Finite metadata still survives selection, final site building, and JavaScript JSON parsing. Both independent reviewers reproduced the failure before the fix and verified rejection afterward; all 164 focused tests passed.

**Review by Arsene12358 — COMMENTED** [review]
Codex whole-branch re-review of `01c009971c796d50d6b3b0ff2c1c612c8b9db559` is complete. The two reported lifecycle defects and the subsequent JSON-overflow finding are fixed. Independent reviewers checked the entire branch against `be8f69749589aea5a376083436e054d75810d379`, reproduced failures themselves, and re-reviewed each correction. No remaining or deferred actionable findings.

## Standards

0 findings. Actual tested provenance, fail-closed malformed input, partial-retry behavior, and trusted site content satisfy the documented contracts. The original `1e999` case now raises before output creation; finite extremes still parse in the final JavaScript consumer.

## Spec

0 findings. Manual dispatches rank their tested main source, older reruns cannot displace newer data, and an unsuccessful retry of the winning qualified run stops deployment before writing. The conservative expired-artifact availability tradeoff is documented.

Both reviewers independently ran the focused Pages/workflow/qualification suite: **164 passed**, including 36 Pages tests. Additional lifecycle checks use the actual qualifier and CSV producer with synthetic probes; real archived FPE data also passes final selection/build/browser rendering (4,134 rows, 78 models, correct provenance, no JavaScript errors).

[Fast CI](https://github.com/ai-dynamo/aisimulate/actions/runs/34943388931), [Pages preview](https://github.com/ai-dynamo/aisimulate/actions/runs/34943388828), DCO, and CODEOWNERS checks passed on this commit. CodeRabbit and Full CI are separate hosted gates; this comment does not assert their completion or replace required human approval.

Summary: Standards 0 (none); Spec 0 (none).

File: scripts/prepare_fpe_pages.py [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+            run.get("path") not in WORKFLOWS
+            or run.get("head_repository", {}).get("full_name") != repository
+            or run.get("head_branch") != "main"
+            or run.get("head_sha") != workflow_sha
+            or run.get("event") not in {"schedule", "workflow_dispatch"}
+        ):
+            continue
+        successful = run.get("status") == "completed" and run.get("conclusion") == "success"
+        if not successful and run.get("run_attempt", 1) <= 1:
+            continue
```
[the remark this answers is not in the document]
Comment (Arsene12358) [reply scripts/prepare_fpe_pages.py:184]: Fixed in c69adb8b60221f79bc3cc1879a8bd57b8f61a1e3. The selector now requires a positive integer `run_attempt` after trusted-producer checks and before any status short-circuit. Missing values, booleans, nonpositive integers, floats, and strings raise ValueError before output creation. The new tests cover 44 malformed-metadata/status combinations with an older valid candidate already selected, and preserve successful retries. Two independent whole-branch reviewers reverified the original failure and the correction against real Git ancestry, including valid retries reaching the site builder. Both reviews are clean; all 167 focused tests (39 Pages cases) passed. This adds the missing test shape for decision-bearing API metadata without changing unrelated API fields.

**Review by Arsene12358 — COMMENTED** [review]
Codex whole-branch re-review is complete at `c69adb8b60221f79bc3cc1879a8bd57b8f61a1e3` after the CodeRabbit run-attempt finding was fixed. The full branch and correction were independently reviewed against `be8f69749589aea5a376083436e054d75810d379`; no remaining or deferred actionable findings.

## Standards

0 findings. `run_attempt` must be a positive integer before status-based selection. Both reviewers independently confirmed that malformed metadata raises without writing output; valid retries continue to the final site builder.

## Spec

0 findings. Actual tested-source ordering, partial-rerun preservation, qualified provenance, malformed-input rejection, trusted HTML, and the documented expired-artifact tradeoff remain correct.

Both independent reviewers ran **167 focused tests**, including **39 Pages cases**. The new 44-case metadata matrix covers missing values, booleans, nonpositive integers, floats, strings, successful statuses, and incomplete/failed statuses with an older valid fallback already selected. Baseline/current probes used actual Git ancestry; the original missing-attempt rollback reproduced before the fix and is rejected afterward.

The final commit also selected and built the recorded real qualified FPE artifact, and Chrome rendered its 78 model rows and correct timestamp/SHA/CI link with no JavaScript errors. This is publication validation, not a GPU accuracy claim. Hosted Fast CI, CodeRabbit, and Full CI are tracked separately on this final commit; required human approvals remain separate.

Summary: Standards 0 (none); Spec 0 (none).

**Comment by Arsene12358** [comment]
@coderabbitai review

Please review the final commit `c69adb8b60221f79bc3cc1879a8bd57b8f61a1e3`. The run-attempt fix is stable; independent Standards/Spec re-reviews and 167 focused tests pass. This single review completes the current-commit review gate after automatic reviews paused.

**Review by Arsene12358 — APPROVED** [review]
